### PR TITLE
POC Tokens::insertSlices

### PR DESCRIPTION
**Pull request by keradus**
ref #4221
cc @kubawerlos , this is what we discussed in priv

### CODE REVIEW

File: tests/Fixer/FunctionNotation/NativeFunctionInvocationFixerTest.php [hunk 1]
```diff
@@ -257,6 +257,20 @@ public function bar($foo)
...
+',
+            ],
+            'fix multiple calls in single code' => [
+                '<?php
+
+\json_encode($foo);
+\strlen($foo);
+\strlen($foo);
+',
+                '<?php
```
Comment (keradus) [inline tests/Fixer/FunctionNotation/NativeFunctionInvocationFixerTest.php]: This, unchanged `Tokens` collection is size of 20.
The fixer has to add 3 new tokens to this collection.

original code has to run `offsetSet` 45 times, as it has to move all tokens from strlen, then middle strlen, then from json_encode - which means that last strlen was moved 3 times!
my proposal has to run `offsetSet` only 20 times, as it has to move one token no more than once

**Comment by keradus** [comment]
why I believe this is a good idea?
I pushed a fix for #3996 to this PR
see the time difference without and with this PR:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3996#issuecomment-455617637

File: tests/AutoReview/FixerTest.php [hunk 2]
```diff
@@ -215,6 +216,79 @@ public function testDeprecatedFixersHaveCorrectSummary(FixerInterface $fixer)
...
 
+    /**
+     * Blur filter that find candidate fixer for performance optimization to use only `insertSlices` instead of multiple `insertAt` if there is no other collection manipulation.
+     *
+     * @dataProvider provideFixerDefinitionsCases
+     */
+    public function testFixerUseInsertSlicesWhenOnlyInsertionsArePerformed(FixerInterface $fixer)
+    {
+        $reflection = new \ReflectionClass($fixer);
+        $tokens = Tokens::froMCode(file_get_contents($reflection->getFileName()));
```
Comment (dmvdbrugge) [inline tests/AutoReview/FixerTest.php]: Typo: uppercase `M` here 😉

Comment (keradus) [reply tests/AutoReview/FixerTest.php]: nice catch :+1:

File: src/Fixer/ArrayNotation/WhitespaceAfterCommaInArrayFixer.php [hunk 3]
```diff
@@ -48,35 +48,32 @@ public function isCandidate(Tokens $tokens)
...
-                $tokens->insertAt($i + 1, new Token([T_WHITESPACE, ' ']));
+            for ($i = $endIndex - 1; $i > $startIndex; --$i) {
+                $i = $this->skipNonArrayElements($i, $tokens);
+                if ($tokens[$i]->equals(',') && !$tokens[$i + 1]->isWhitespace()) {
+                    $tokensToInsert[$i + 1] = new Token([T_WHITESPACE, ' ']);
+                }
             }
         }
+
+        if ($tokensToInsert) {
```
Comment (kubawerlos) [inline src/Fixer/ArrayNotation/WhitespaceAfterCommaInArrayFixer.php]: `[] === $tokensToInsert`?

File: src/Tokenizer/Tokens.php [hunk 4]
```diff
@@ -869,7 +869,27 @@ public function findSequence(array $sequence, $start = 0, $end = null, $caseSens
     public function insertAt($index, $items)
     {
         $items = \is_array($items) || $items instanceof self ? $items : [$items];
-        $itemsCnt = \count($items);
+
+        $this->insertSlices([$index => $items]);
+    }
+
+    /**
+     * @TODO: docs
```
Comment (kubawerlos) [inline src/Tokenizer/Tokens.php]: do?

File: src/Tokenizer/Tokens.php [hunk 5]
```diff
@@ -869,7 +869,27 @@ public function findSequence(array $sequence, $start = 0, $end = null, $caseSens
...
+     *
+     * @internal
+     *
+     * 5 => tokenA
+     * 7 => [tokenB, tokenC]
+     *
+     * @param mixed $slices
```
Comment (kubawerlos) [inline src/Tokenizer/Tokens.php]: `array<int, Token>`?

File: src/Tokenizer/Tokens.php [hunk 6]
```diff
@@ -869,7 +869,27 @@ public function findSequence(array $sequence, $start = 0, $end = null, $caseSens
...
+     */
+    public function insertSlices($slices)
+    {
+        $itemsCnt = 0;
```
Comment (kubawerlos) [inline src/Tokenizer/Tokens.php]: `$itemsCount` (or `$itemCount`) - the point is not to abbreviate.

File: src/Tokenizer/Tokens.php (hunk 5 above)
Comment (dmvdbrugge) [reply src/Tokenizer/Tokens.php]: If I look at the description it should be `array<int, Token|Token[]>`

Comment (keradus) [reply src/Tokenizer/Tokens.php]: indeed, i simply didn't documented this methods (as I put todo), as i wanted the idea to be validated first ;)

File: src/Fixer/ArrayNotation/WhitespaceAfterCommaInArrayFixer.php (hunk 3 above)
Comment (dmvdbrugge) [reply src/Fixer/ArrayNotation/WhitespaceAfterCommaInArrayFixer.php]: You mean `!==`

Comment (keradus) [reply src/Fixer/ArrayNotation/WhitespaceAfterCommaInArrayFixer.php]: done

File: src/Tokenizer/Tokens.php (hunk 6 above)
Comment (keradus) [reply src/Tokenizer/Tokens.php]: done

File: src/Tokenizer/Tokens.php (hunk 4 above)
Comment (keradus) [reply src/Tokenizer/Tokens.php]: done

File: src/Tokenizer/Tokens.php [hunk 7]
```diff
@@ -869,31 +869,61 @@ public function findSequence(array $sequence, $start = 0, $end = null, $caseSens
...
 
         // since we only move already existing items around, we directly call into SplFixedArray::offset* methods.
         // that way we get around additional overhead this class adds with overridden offset* methods.
-        for ($i = $oldSize + $itemsCnt - 1; $i >= $index; --$i) {
-            $oldItem = parent::offsetExists($i - $itemsCnt) ? parent::offsetGet($i - $itemsCnt) : new Token('');
-            parent::offsetSet($i, $oldItem);
-        }
+        foreach ($slices as $index => $slice) {
+            $slice = \is_array($slice) || $slice instanceof self ? $slice : [$slice];
+            $sliceCnt = \count($slice);
```
Comment (dmvdbrugge) [inline src/Tokenizer/Tokens.php]: Nonblocking: I guess this should become `sliceCount` as well, given @kubawerlos's reasoning

File: src/Tokenizer/Tokens.php [hunk 8]
```diff
@@ -869,31 +869,61 @@ public function findSequence(array $sequence, $start = 0, $end = null, $caseSens
...
-        if (0 === $itemsCnt) {
+        $this->insertSlices([$index => $items]);
+    }
+
+    /**
+     * Insert a slices or individual Tokens into multiple places in a single run.
+     *
+     * @internal
+     *
+     * @param array<int, Token|Token[]> $slices
```
Comment (dmvdbrugge) [inline src/Tokenizer/Tokens.php]: What I'm missing in this doc is the fact that this `int` key is used as insert-index, while it is mentioned in for example `insertAt`. Leaving it out here may cause one to think that it's not the same idea.

Also, isn't `Tokens` also accepted, thus shouldn't the typehint be `array<int, Token|Token[]|Tokens>`?

**Comment by keradus** [comment]
@julienfalque , @SpacePossum , what do you think about the proposal in this PR? especially do we want to promote new method as new "best practice" or not?
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4250/files#diff-0ae0077906c99aaf35085e05c46cefd3R284

> I think the issue is not related to the overall length, so all keys/values can be probably changed to single char like `0` (zero) or `''` (empty string).

it can, i was just too lazy to modify example provided by original reporter.

File: src/Tokenizer/Tokens.php (hunk 8 above)
Comment (keradus) [reply src/Tokenizer/Tokens.php:907]: :+1:

**Comment by keradus** [comment]
> Is the name `insertSlices` fitting situation? Quite often (if not most times) we will use this to insert array of tokens. Wouldn't `insertMultiple` (or `insertMany`) better describe what the function is doing?

for me, `insertMultiple`/`insertMany` is more fitting actual behaviour of `insertAt`, where we are pointing a single index, and adding multiple tokens to it:
```php
$tokens->insertAt(15, $tokensCollection)
$tokens->insertAt(10, [$tokenA, $tokenB]);
$tokens->insertAt(5, $tokenC);
```

I'm trying to express that we are inserting different slices to different indexes at single call:
```php
$tokens->insertSlices([
    5 => $tokenC,
    10 => [$tokenA, $tokenB],
    15 => $tokensCollection,
]);
```

**Comment by mvorisek** [comment]
I was hoping this PR with solve https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5279, but even with this PR running CS Fixer against https://github.com/mvorisek/TCPDF/tree/psr12_cs is slow and CS Fixer does not finish within an hour.

**Comment by keradus** [comment]
@mvorisek :
- please avoid thread hijacking
- I already shared that this PR is NOT helping with topic you raised - https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5279#issuecomment-740656516
- if you want your edge-case topic covered, consider the suggestion raised here: https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/4702#issuecomment-740749024

**Pull request merged**